# Peer-graded-assignemt: Segmenting and Clustering Neighborhoods in Toronto

## Part 1

Importing urllib and downloading the html data from Wikipedia. <br>
Then importing BeautifulSoup and creating a table from that data:

In [1]:
import urllib.request

url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
page = urllib.request.urlopen(url)

from bs4 import BeautifulSoup
soup = BeautifulSoup(page, "lxml")
right_table=soup.find('table', class_='wikitable sortable')

Create three empty lists for the three columns 'PostalCodes', 'Boroughs' and 'Neighborhoods'<br>
From the BeautifulSoup table find the entries with the 'tr' tag, from those find the entries with the 'td' tag and assign them to the variable cells <br>
If there are three entries in the variable cells (one for PostalCodes, one for Boroughs and one for Neighborhoods), take the text, turn it into a string, strip the linebreak charachter and append them to the corresponding list:

In [2]:
PostalCodes = []
Boroughs = []
Neighborhoods = []

for row in right_table.findAll('tr'):
    cells = row.findAll('td')
    if len(cells) == 3:
        PostalCodes.append(str(cells[0].find(text=True)).strip("\n"))
        Boroughs.append(str(cells[1].find(text=True)).strip("\n"))
        Neighborhoods.append(str(cells[2].find(text=True)).strip("\n"))


import Pandas and create a dataframe with three columns. The Column names are 'PostalCode', 'Borough' and 'Neighborhood'<br>
Fill the columns of the dataframe with the entries from the list of PostalCodes, Boroughs and Neighborhoods <br>
Print the first five entries to verify the data:

In [3]:
import pandas as pd
df=pd.DataFrame(PostalCodes,columns=['PostalCode'])
df['Borough']=Boroughs
df['Neighborhood']= Neighborhoods

df.head()

PostalCode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

Loop through every row of the Dataframe and create a list of all the rows where the value for Borough is 'Not assigned'.<br>
Then remove all the rows from the DataFrame and print out the first five rows to verify the data

In [4]:
i = 0
index_list = []

for element in df['Borough']:
    if str(df.loc[i, 'Borough']).strip() == 'Not assigned':
        index_list.append(i)
        i += 1
    else:
        i += 1

df_mod = df.drop(df.index[index_list])
df_mod.head()

PostalCode           Borough                                 Neighborhood
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Since the index does not match the row numbers anymore, reset the index. 

In [5]:
df_mod = df_mod.reset_index(drop=True)
df_mod.head()

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Print out the shape of the DataFrame as instructed:

In [6]:
df_mod.shape

(103, 3)

## Part 2:

Create a DataFrame with the coordinates of the different Postal Codes in Toronto <br>
Note: Since Geocoder is unreliable I decided to download the .csv file with the postal codes instead of using Geocoder

In [7]:
file = (pd.read_csv('Geospatial_Coordinates.csv'))
coords = pd.DataFrame(file)
coords

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

Renaming the Column "Postal Code" to "PostalCode" to match the DataFrame from part 1:

In [8]:
coords.rename(columns={"Postal Code": "PostalCode"}, inplace = True)

Merging the two DataFrames on PostalCode:

In [9]:
df_complete = df_mod.merge(coords, on = "PostalCode")

Displaying the first 12 entries of the merged DataFrame:

In [10]:
df_complete.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

## Part 3:

Defining foursquare credentials and Version:<br>
Note: I have changed the credentials after running the code so that no one can use my Foursquare credentials

In [11]:
CLIENT_ID = 'ESKBE3EZFPO2QMCBJSNX4QTUWIK2PR0EQAK4243DDKM45B3F'
CLIENT_SECRET = 'EIWEA3UT3P0KUKJK52SXBLG2UEFYO4L0TQAWIARV0VT2GNRD'
VERSION = '20180605' 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ESKBE3EZFPO2QMCBJSNX4QTUWIK2PR0EQAK4243DDKM45B3F
CLIENT_SECRET:EIWEA3UT3P0KUKJK52SXBLG2UEFYO4L0TQAWIARV0VT2GNRD


Used the code from the prevoius lab to get a list of venues for the Neighborhoods in our DataFrame

In [12]:
import requests
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT = 100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Run the function with the parameters from our Toronto DataFrame:

In [13]:
toronto_venues = getNearbyVenues(names=df_complete['Neighborhood'],
                                   latitudes=df_complete['Latitude'],
                                   longitudes=df_complete['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Check the output of the api call to see if the venues have been added correctly

In [14]:
print(toronto_venues.shape)
toronto_venues.head(10)

(2140, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                  Parkwoods              43.753259              -79.329656   
1                  Parkwoods              43.753259              -79.329656   
2                  Parkwoods              43.753259              -79.329656   
3           Victoria Village              43.725882              -79.315572   
4           Victoria Village              43.725882              -79.315572   
5           Victoria Village              43.725882              -79.315572   
6           Victoria Village              43.725882              -79.315572   
7           Victoria Village              43.725882              -79.315572   
8           Victoria Village              43.725882              -79.315572   
9  Regent Park, Harbourfront              43.654260              -79.360636   

                                       Venue  Venue Latitude  Venue Longitude  \
0                            Brookbanks Park       43.751976       -79.332140   
1                            Brookbanks Pool       43.751389       -79.332184   
2                              Variety Store       43.751974       -79.333114   
3                     Victoria Village Arena       43.723481       -79.315635   
4                                  Portugril       43.725819       -79.312785   
5                                Tim Hortons       43.725517       -79.313103   
6                                   The Frig       43.727051       -79.317418   
7  Eglinton Ave E & Sloane Ave/Bermondsey Rd       43.726086       -79.313620   
8                                 Pizza Nova       43.725824       -79.312860   
9                           Roselle Desserts       43.653447       -79.362017   

          Venue Category  
0                   Park  
1                   Pool  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant  
5            Coffee Shop  
6      French Restaurant  
7           Intersection  
8            Pizza Place  
9                 Bakery

Count the number of entries for each neighborhood:

In [15]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                               8   
Bathurst Manor, Wilson Heights, Downsview North                     20   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   26   
...                                                                ...   
Willowdale, Willowdale West                                          7   
Woburn                                                               4   
Woodbine Heights                                                     8   
York Mills West                                                      2   
York Mills, Silver Hills                                             1   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                                8   
Bathurst Manor, Wilson Heights, Downsview North                      20   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    26   
...                                                                 ...   
Willowdale, Willowdale West                                           7   
Woburn                                                                4   
Woodbine Heights                                                      8   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                               8               8   
Bathurst Manor, Wilson Heights, Downsview North     20              20   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   26              26   
...                                                ...             ...   
Willowdale, Willowdale West                          7               7   
Woburn                                               4               4   
Woodbine Heights                                     8               8   
York Mills West                                      2               2   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                         8   
Bathurst Manor, Wilson Heights, Downsview North               20   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             26   
...                                                          ...   
Willowdale, Willowdale West                                    7   
Woburn                                                         4   
Woodbine Heights                                               8   
York Mills West                                                2   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                        8 

Check the number of unique venue categories:

In [16]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 272 uniques categories.


Apply one-hot-encoding and create a new dataframe with the encoded value for each venue per neighborhood, displaying the venue category in the column names:

In [17]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
toronto_onehot = toronto_onehot[ ['Neighborhood'] + [ col for col in toronto_onehot.columns if col != 'Neighborhood' ] ]
toronto_onehot

Neighborhood  Accessories Store  \
0                                             Parkwoods                  0   
1                                             Parkwoods                  0   
2                                             Parkwoods                  0   
3                                      Victoria Village                  0   
4                                      Victoria Village                  0   
...                                                 ...                ...   
2135  Mimico NW, The Queensway West, South of Bloor,...                  0   
2136  Mimico NW, The Queensway West, South of Bloor,...                  0   
2137  Mimico NW, The Queensway West, South of Bloor,...                  0   
2138  Mimico NW, The Queensway West, South of Bloor,...                  0   
2139  Mimico NW, The Queensway West, South of Bloor,...                  0   

      Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0                     0        0                   0               0   
1                     0        0                   0               0   
2                     0        0                   0               0   
3                     0        0                   0               0   
4                     0        0                   0               0   
...                 ...      ...                 ...             ...   
2135                  0        0                   0               0   
2136                  0        0                   0               0   
2137                  0        0                   0               0   
2138                  0        0                   0               0   
2139                  0        0                   0               0   

      Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                   0                 0                    0             0   
1                   0                 0                    0             0   
2                   0                 0                    0             0   
3                   0                 0                    0             0   
4                   0                 0                    0             0   
...               ...               ...                  ...           ...   
2135                0                 0                    0             0   
2136                0                 0                    0             0   
2137                0                 0                    0             0   
2138                0                 0                    0             0   
2139                0                 0                    0             0   

      ...  Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0     ...                              0                 0            0   
1     ...                              0                 0            0   
2     ...                              0                 0            0   
3     ...                              0                 0            0   
4     ...                              0                 0            0   
...   ...                            ...               ...          ...   
2135  ...                              0                 0            0   
2136  ...                              0                 0            0   
2137  ...                              0                 0            0   
2138  ...                              0                 0            0   
2139  ...                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  \
0                         0                0         0          0   
1                         0                0         0          0   
2                         0                0         0          0   
3                         0                0         0          0   
4                         0                0         0    

Check the shape of the new DataFrame to see that there are 272 unique columns for the unique venue types now:

In [18]:
toronto_onehot.shape

(2140, 272)

Group the results by neighborhood, get the mean and reset the index:

In [19]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
90                      Willowdale, Willowdale West                0.0   
91                                           Woburn                0.0   
92                                 Woodbine Heights                0.0   
93                                  York Mills West                0.0   
94                         York Mills, Silver Hills                0.0   

    Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0                 0.0      0.0                 0.0             0.0   
1                 0.0      0.0                 0.0             0.0   
2                 0.0      0.0                 0.0             0.0   
3                 0.0      0.0                 0.0             0.0   
4                 0.0      0.0                 0.0             0.0   
..                ...      ...                 ...             ...   
90                0.0      0.0                 0.0             0.0   
91                0.0      0.0                 0.0             0.0   
92                0.0      0.0                 0.0             0.0   
93                0.0      0.0                 0.0             0.0   
94                0.0      0.0                 0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0               0.0             0.000000           0.0  ...   
1               0.0               0.0             0.000000           0.0  ...   
2               0.0               0.0             0.000000           0.0  ...   
3               0.0               0.0             0.000000           0.0  ...   
4               0.0               0.0             0.038462           0.0  ...   
..              ...               ...                  ...           ...  ...   
90              0.0               0.0             0.000000           0.0  ...   
91              0.0               0.0             0.000000           0.0  ...   
92              0.0               0.0             0.000000           0.0  ...   
93              0.0               0.0             0.000000           0.0  ...   
94              0.0               0.0             0.000000           0.0  ...   

    Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                             0.0               0.0          0.0   
1                             0.0               0.0          0.0   
2                             0.0               0.0          0.0   
3                             0.0               0.0          0.0   
4                             0.0               0.0          0.0   
..                            ...               ...          ...   
90                            0.0               0.0          0.0   
91                            0.0               0.0          0.0   
92                            0.0               0.0          0.0   
93                            0.0               0.0          0.0   
94                            0.0               0.0          0.0   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                     0.0              0.0       0.0        0.0          0.0   
1                     0.0              0.0       0.0        0.0          0.0   
2                     0.0              0.0       0.0        0.0          0.0   
3                     0.0              0.0       0.0        0.0          0.0   
4                     0.0              0.0       0.0        0.0          0.0   
..                    ...              

Function to sort venues in descending order:

In [20]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create a new dataframe with the name of the neighborhood and adding a ranking of the most common venues by calling the function return_most_common_venues for each row in the toronto_grouped DataFrame:

In [21]:
import numpy as np
num_top_venues = 10

columns = ['Neighborhood']
columns.append('1st Most Common Venue')
columns.append('2nd Most Common Venue')
columns.append('3rd Most Common Venue')
columns.append('4th Most Common Venue')
columns.append('5th Most Common Venue')
columns.append('6th Most Common Venue')
columns.append('7th Most Common Venue')
columns.append('8th Most Common Venue')
columns.append('9th Most Common Venue')
columns.append('10th Most Common Venue')

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0          Skating Rink  Latin American Restaurant        Breakfast Spot   
1                   Gym                       Pool           Coffee Shop   
2           Coffee Shop              Shopping Mall           Pizza Place   
3                  Bank        Japanese Restaurant    Chinese Restaurant   
4        Sandwich Place                 Restaurant           Coffee Shop   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0             Drugstore             Discount Store   Distribution Center   
1        Sandwich Place         Athletics & Sports                   Pub   
2      Sushi Restaurant  Middle Eastern Restaurant            Restaurant   
3           Yoga Studio             Discount Store   Distribution Center   
4           Pizza Place    Comfort Food Restaurant       Thai Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dog Run      Doner Restaurant             Donut Shop  
1    Dim Sum Restaurant                 Diner         Discount Store  
2        Ice Cream Shop     Mobile Phone Shop    Fried Chicken Joint  
3               Dog Run      Doner Restaurant             Donut Shop  
4             Juice Bar                   Spa                Butcher

import KMeans to cluster the toronto_grouped DataFrame. <br>
Drop the Neighborhood column (otherwise KMeans will fail), create the KMeans model with 5 clusters and fit it. <br>
Afterwards, return the created labels:

In [22]:
from sklearn.cluster import KMeans
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

kmeans.labels_[0:] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 4, 0, 1, 3, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 0, 1])

Insert the cluster labels into the neighborhood_venues_sorted dataframe and create a new dataframe in which the Dataframe from part 2 (with columns PostalCode,	Borough, Neighborhood, Latitude, Longitude) and the dataframe neighborhoods_venues_sorted are merged. <br>
Then drop all rows without values in the 'Cluster Labels' cloumn and display the first five rows of the new dataframe

In [23]:
try:
    neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
except:
    pass

toronto_merged = df_complete

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
toronto_merged.dropna(0, inplace = True)
toronto_merged.head()

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels  1st Most Common Venue  \
0  43.753259 -79.329656             1.0                   Park   
1  43.725882 -79.315572             1.0  Portuguese Restaurant   
2  43.654260 -79.360636             1.0            Coffee Shop   
3  43.718518 -79.464763             1.0         Clothing Store   
4  43.662301 -79.389494             1.0            Coffee Shop   

  2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop                    Pool    Dim Sum Restaurant   
1     French Restaurant             Pizza Place           Coffee Shop   
2                Bakery                     Pub                  Park   
3     Accessories Store  Furniture / Home Store           Event Space   
4                 Diner        Sushi Restaurant           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Diner        Discount Store   Distribution Center   
1           Hockey Arena          Intersection           Yoga Studio   
2         Breakfast Spot                  Café               Theater   
3  Vietnamese Restaurant           Coffee Shop              Boutique   
4                   Park                   Bar              Beer Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dog Run      Doner Restaurant             Donut Shop  
1          Dessert Shop    Dim Sum Restaurant                  Diner  
2            Beer Store            Shoe Store             Restaurant  
3    Miscellaneous Shop               Dog Run     Dim Sum Restaurant  
4         Smoothie Shop        Sandwich Place          Burrito Place

Import folium to display the datapoints from the toronto_merged dataframe on a map of Toronto:

In [26]:
import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

map_clusters = folium.Map(location=[43.7000, -79.4000], zoom_start=11)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Note: The map doesn't show on github but works outside of it